crates/noirc_printable_type/src/lib.rs

```diff
@@ -0,0 +1,199 @@
+    }
+}
+
+impl<'a> std::fmt::Display for PrintableValueDisplay<'a> {
+    fn fmt(&self, fmt: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match (&self.value, &self.typ) {
+            (
+                PrintableValue::Field(f),
+                PrintableType::Field
+                // TODO: We should print the sign for these and probably print normal integers instead of field strings
+                | PrintableType::SignedInteger { .. }
+                | PrintableType::UnsignedInteger { .. },
+            ) => {
+                write!(fmt, "{}", format_field_string(*f))?;
+            }
+            (PrintableValue::Field(f), PrintableType::Boolean) => {
+                if f.is_one() {
+                    write!(fmt, "true")?;
+                } else {
+                    write!(fmt, "false")?;
+                }
+            }
+            (PrintableValue::Vec(vector), PrintableType::Array { typ, .. }) => {
+                write!(fmt, "[")?;
+                let mut values = vector.iter().peekable();
+                while let Some(value) = values.next()  {
+                    write!(fmt, "{}", PrintableValueDisplay::new(value, typ))?;
+                    if values.peek().is_some() {
+                        write!(fmt, ", ")?;
+                    }
+                }
+                write!(fmt, "]")?;
+            }
+
+            (PrintableValue::String(s), PrintableType::String { .. }) => {
+                write!(fmt, r#""{s}""#)?;
+            }
+
+            (PrintableValue::Struct(map), PrintableType::Struct { name, fields, .. }) => {
+                write!(fmt, "{name} {{ ")?;
+
+                let mut fields = fields.iter().peekable();
+                while let Some((key, field_type)) = fields.next()  {
+                    let value = &map[key];
+                    write!(fmt, "{key}: {}", PrintableValueDisplay::new(value, field_type))?;
+                    if fields.peek().is_some() {
+                        write!(fmt, ", ")?;
+                    }
+                }
+
+                write!(fmt, " }}")?;
+            }
+
+            _ => return Err(std::fmt::Error),
+        };
+        Ok(())
+    }
+}
+
+/// This trims any leading zeroes.
+/// A singular '0' will be prepended as well if the trimmed string has an odd length.
+/// A hex string's length needs to be even to decode into bytes, as two digits correspond to
+/// one byte.
+fn format_field_string(field: FieldElement) -> String {
+    if field.is_zero() {
+        return "0x00".to_owned();
+    }
+    let mut trimmed_field = field.to_hex().trim_start_matches('0').to_owned();
+    if trimmed_field.len() % 2 != 0 {
+        trimmed_field = "0".to_owned() + &trimmed_field;
+    }
+    "0x".to_owned() + &trimmed_field
+}
+
+// TODO: Figure out a better API for this to avoid exporting the function
+/// Assumes that `field_iterator` contains enough [FieldElement] in order to decode the [PrintableType]
+pub fn decode_value(
```

Comment:
Could we instead have a `from_values` method of some kind on `PrintableValueDisplay` that takes in a `PrintableType` and list of values? The `from` method then uses `decode_value` to construct a `PrintableValue`.

Reply:
@vezenovm I just did a big refactor so `PrintableValueDisplay` could implement `TryFrom<&[Vec<Value>]>`. Please take a look.